include/deal.II/sundials/n_vector.h

```diff
@@ -0,0 +1,116 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2020 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE.md at
+//    the top level directory of deal.II.
+//
+//-----------------------------------------------------------
+
+#ifndef dealii_sundials_n_vector_h
+#define dealii_sundials_n_vector_h
+
+#include <deal.II/base/config.h>
+
+#if defined(DEAL_II_WITH_SUNDIALS) && DEAL_II_SUNDIALS_VERSION_GTE(5, 0, 0)
+#  include <sundials/sundials_nvector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace SUNDIALS
+{
+  namespace internal
+  {
+    /**
+     * Retrieve the underlying vector attached to N_Vector @p v. This call will
+     * only succeed if the underlying vector is not const. Use
+     * unwrap_nvector_const() for this case.
+     *
+     * @note Users must ensure that they ask for the correct VectorType when
+     *   calling this function and there are no type-safety checks in place.
+     *
+     * @tparam VectorType type of the vector that is stored in @p v
+     * @param v vector to unwrap
+     * @return the vector that is stored inside @p v
+     */
+    template <typename VectorType>
+    VectorType *
+    unwrap_nvector(N_Vector v);
+
+    /**
+     * Retrieve the underlying vector attached to N_Vector @p v as a constant
+     * pointer.
+     *
+     * @note Users must ensure that they ask for the correct VectorType when
+     *   calling this function and there are no type-safety checks in place.
+     *
+     * @tparam VectorType type of the vector that is stored in @p v
+     * @param v vector to unwrap
+     * @return the vector that is stored inside @p v
+     */
+    template <typename VectorType>
+    const VectorType *
+    unwrap_nvector_const(N_Vector v);
```

Comment:
Not sure if the calls should explicitly mention the const-ness. Alternative would be to enable something like `unwrap_nvector<const VectorType>(...)` but this feels unintuitive to me.